File: android/src/main/kotlin/com/nover/flutternativeadmob/NativeAdmobController.kt
package com.nover.flutternativeadmob

import android.content.Context
import android.util.Log
import com.google.android.gms.ads.AdListener
import com.google.android.gms.ads.AdLoader
import com.google.android.gms.ads.AdRequest
import com.google.android.gms.ads.formats.UnifiedNativeAd
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class NativeAdmobController(
    val id: String,
    private val channel: MethodChannel,
    private val context: Context
) : MethodChannel.MethodCallHandler {

  enum class CallMethod {
    initAd, reloadAd
  }

  enum class LoadState {
    loading, loadError, loadCompleted
  }

  var nativeAdChanged: ((UnifiedNativeAd?) -> Unit)? = null
  var nativeAd: UnifiedNativeAd? = null
    set(value) {
      field = value
      invokeLoadCompleted()
    }

  private var adLoader: AdLoader? = null
  private var adUnitID: String? = null

  init {
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
    when (CallMethod.valueOf(call.method)) {
      CallMethod.initAd -> {
        call.argument<String>("adUnitID")?.let {
          val isChanged = adUnitID != it
          adUnitID = it

          if (adLoader == null || isChanged) {
            val builder = AdLoader.Builder(context, it)
            adLoader = builder.forUnifiedNativeAd { nativeAd ->
              this.nativeAd = nativeAd
            }.withAdListener(object : AdListener() {
              override fun onAdFailedToLoad(errorCode: Int) {
                println("onAdFailedToLoad errorCode = $errorCode")
                channel.invokeMethod(LoadState.loadError.toString(), null)
              }
            }).build()
          }

          if (nativeAd == null || isChanged) loadAd() else invokeLoadCompleted()
        } ?: result.success(null)
      }

      CallMethod.reloadAd -> {
        call.argument<Boolean>("forceRefresh")?.let {
          if (it || nativeAd == null) loadAd() else invokeLoadCompleted()
        }
      }
    }
  }

  private fun loadAd() {
    channel.invokeMethod(LoadState.loading.toString(), null)
    adLoader?.loadAd(AdRequest.Builder().build())
  }

  private fun invokeLoadCompleted() {
    nativeAdChanged?.let { it(nativeAd) }
    channel.invokeMethod(LoadState.loadCompleted.toString(), null)
  }
}

object NativeAdmobControllerManager {
  private val controllers: ArrayList<NativeAdmobController> = arrayListOf()

  fun createController(id: String, binaryMessenger: BinaryMessenger, context: Context) {
    if (getController(id) == null) {
      val methodChannel = MethodChannel(binaryMessenger, id)
      val controller = NativeAdmobController(id, methodChannel, context)
      controllers.add(controller)
    }
  }

  fun getController(id: String): NativeAdmobController? {
    return controllers.firstOrNull { it.id == id }
  }

  fun removeController(id: String) {
    val index = controllers.indexOfFirst { it.id == id }
    if (index >= 0) controllers.removeAt(index)
  }
}

File: ios/Classes/NativeAdmobController.swift
//
//  NativeAdmobController.swift
//  flutter_native_admob
//
//  Created by Dao Duy Duong on 3/14/20.
//

import GoogleMobileAds

class NativeAdmobController: NSObject {
    
    enum CallMethod: String {
        case setAdUnitID
        case reloadAd
        case setNonPersonalizedAds
    }
    
    enum LoadState: String {
        case loading, loadError, loadCompleted
    }
    
    let id: String
    let channel: FlutterMethodChannel
    
    var nativeAdChanged: ((GADUnifiedNativeAd?) -> Void)?
    var nativeAd: GADUnifiedNativeAd? {
        didSet { invokeLoadCompleted() }
    }
    
    private var adLoader: GADAdLoader?
    private var adUnitID: String?
    private var nonPersonalizedAds: Bool = false
    
    init(id: String, channel: FlutterMethodChannel) {
        self.id = id
        self.channel = channel
        super.init()
        
        channel.setMethodCallHandler(handle)
    }
    
    private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let callMethod = CallMethod(rawValue: call.method) else { return result(FlutterMethodNotImplemented) }
        let params = call.arguments as? [String: Any]
        
        switch callMethod {
        case .setAdUnitID:
            guard let adUnitID = params?["adUnitID"] as? String else {
                return result(nil)
            }
            let isChanged = adUnitID != self.adUnitID
            self.adUnitID = adUnitID
            if adLoader == nil || isChanged {
                let numberAds: Int = params?["numberAds"] as? Int ?? 1
                let multipleAdsOptions = GADMultipleAdsAdLoaderOptions()
                if numberAds != nil && numberAds > 1 {
                    multipleAdsOptions.numberOfAds = numberAds
                }
                adLoader = GADAdLoader(
                    adUnitID: adUnitID, 
                    rootViewController: nil, 
                    adTypes: [.unifiedNative], 
                    options: [multipleAdsOptions]
                )
                adLoader?.delegate = self
            }
            if nativeAd == nil || isChanged {
                loadAd()
            } else {
                invokeLoadCompleted()
            }
            
        case .reloadAd:
            let forceRefresh = params?["forceRefresh"] as? Bool ?? false
            if forceRefresh || nativeAd == nil {
                loadAd()
            } else {
                invokeLoadCompleted()
            }
            
        case .setNonPersonalizedAds:
            if let nonPersonalizedAds = params?["nonPersonalizedAds"] as? Bool {
                self.nonPersonalizedAds = nonPersonalizedAds
            }
            return result(nil)
        }
        
        result(nil)
    }
    
    private func loadAd() {
        channel.invokeMethod(LoadState.loading.rawValue, arguments: nil)
        let request = GADRequest()
        if self.nonPersonalizedAds {
            let extras = GADExtras()
            extras.additionalParameters = ["npa": "1"]
            request.register(extras)
        }
        adLoader?.load(request)
    }
    
    private func invokeLoadCompleted() {
        nativeAdChanged?(nativeAd)
        channel.invokeMethod(LoadState.loadCompleted.rawValue, arguments: nil)
    }
}

extension NativeAdmobController: GADUnifiedNativeAdLoaderDelegate {
    
    func adLoader(_ adLoader: GADAdLoader, didFailToReceiveAdWithError error: GADRequestError) {
        print("NativeAdmob: failed to load with error: \(error.localizedDescription)")
        channel.invokeMethod(LoadState.loadError.rawValue, arguments: nil)
    }
    
    func adLoader(_ adLoader: GADAdLoader, didReceive nativeAd: GADUnifiedNativeAd) {
        self.nativeAd = nativeAd
    }
}

class NativeAdmobControllerManager {
    
    static let shared = NativeAdmobControllerManager()
    
    private var controllers: [NativeAdmobController] = []
    
    private init() {}
    
    func createController(forID id: String, binaryMessenger: FlutterBinaryMessenger) {
        if getController(forID: id) == nil {
            let methodChannel = FlutterMethodChannel(name: id, binaryMessenger: binaryMessenger)
            let controller = NativeAdmobController(id: id, channel: methodChannel)
            controllers.append(controller)
        }
    }
    
    func getController(forID id: String) -> NativeAdmobController? {
        return controllers.first(where: { $0.id == id })
    }
    
    func removeController(forID id: String) {
        if let index = controllers.firstIndex(where: { $0.id == id }) {
            controllers.remove(at: index)
        }
    }
}
